packages/discovery-provider/src/tasks/index_payment_router.py

```diff
@@ -333,60 +358,232 @@ def index_purchase(
+        )
+
+
+def attempt_index_recovery_transfer(
+    session: Session,
+    sender_account: str,
+    sender_user_account: UserIdBankAccount | None,
+    receiver_user_accounts: List[UserIdBankAccount],
+    balance_changes: dict[str, BalanceChange],
+    tx_sig: str,
+):
+    if sender_user_account is not None:
+        raise Exception(
+            f"Sender account for recovery cannot be a user bank: {sender_user_account}"
+        )
+    if len(receiver_user_accounts) != 1:
+        raise Exception(
+            f"Recovery transfer must have exactly one receiver account. Received: {','.join([a['user_bank_account'] for a in receiver_user_accounts])}"
+        )
+
+    receiver_user_account = receiver_user_accounts[0]
+    receiver_bank_account = receiver_user_account["user_bank_account"]
+    receiver_balance_change = balance_changes[receiver_bank_account]
+    receiver_balance_change_amount = receiver_balance_change["change"]
+
+    # Find the previous transaction in the other direction (receiver -> sender)
+    # So that we can modify it
+    previous_transaction = (
+        session.query(USDCTransactionsHistory)
+        .filter(
+            and_(
+                USDCTransactionsHistory.user_bank == receiver_bank_account,
+                USDCTransactionsHistory.method == USDCTransactionMethod.send,
+                USDCTransactionsHistory.tx_metadata == sender_account,
+            )
+        )
+        .order_by(desc(USDCTransactionsHistory.transaction_created_at))
+        .first()
+    )
+    if previous_transaction is None:
+        raise Exception(
+            f"Failed to find matching previous transcation from {receiver_bank_account} to {sender_account}"
+        )
+
+    difference = previous_transaction.change + receiver_balance_change_amount
+
+    # Previous transaction is undone by this recovery
+    if difference == 0:
+        logger.info(
+            f"index_payment_router.py | tx: {tx_sig} | Removing previous transaction {previous_transaction.signature}."
+        )
+        session.delete(previous_transaction)
+    elif difference < 0:
+        # Previous transaction is partially undone by this recovery
+        # Modifying the object returned by the query will update the DB
+        # when the session is flushed
+        previous_transaction.change = (
+            previous_transaction.change + receiver_balance_change_amount
+        )
+        previous_transaction.balance = Decimal(receiver_balance_change["post_balance"])
+        logger.info(
+            f"index_payment_router.py | tx: {tx_sig} | Found partial recovery for  {previous_transaction.signature}. New change value is {previous_transaction.change}."
+        )
+    else:
+        raise Exception(
+            f"Recovery transaction is for greater amount than original transaction ({receiver_balance_change_amount} > {-previous_transaction.change})"
```

Comment:
I'm back and forth on this. The only use cases I can think of where the incoming amount would be greater are if more tokens somehow got transferred into the root wallet token account before recovery happened. In our client, the only way this should happen is if we tried to initiate two partial withdraws in quick succession without successfully recovering.
And in that case, we wouldn't want to just undo the most recent. We would want to loop through the most recent and undo until we've exhausted the recovered funds.

Seems unnecessarily complicated for an edge case. And the way this will read is more like two "holds" that get refunded shortly after.